### DISCUSSION TITLE
OAuthAccountNotLinked for pre-existing users without account

### DISCUSSION THREAD

**Discussion by Gregoor** [nextauthjs/next-auth#3125] [Help]
This is probably as expected, so not a bug per-se, but it interferes with our apps needs. We want to create users before they have an account to email them a session-link (basically a JWT in the URL's query params, which gets put into the cookies on first load) usable for certain actions. While that flow works nicely, and they can then also claim that account by signing in, it fails when they try to sign-in without using the session-link we generated. We're getting an error from here:
https://github.com/nextauthjs/next-auth/blob/86baefdd9df9246669d4f788aaab4774605c256a/src/server/lib/callback-handler.js#L204-L213

In our best case that condition would not just check for the existence of a user by email, but also whether that user has an account. I think that also kind of aligns with the expectation formulated by the comment i.e.
> *BUT* we do already have an account with the same email address

(which is not the case necessarily, as in our case we have a user without an account)

I'd be curious to hear any ideas around this! Thanks

**Comment by kcrtr** [comment]
Is there a specific reason you are not sending temporary passwords?

**Reply by Gregoor** [reply]
We'd want it to be as frictionless as clicking a link, though I guess we could also have that temp password in the URL and programmtically log-in with it. I will look into it, thanks for the pointer!

**Reply by balazsorban44** [reply]
I think you don't want to set passwords in query parameters.

**Reply by kcrtr** [reply]
I was more specifically thinking that you could send the temp password in an email.

**Comment by balazsorban44** [comment]
Check out the Email Provider https://next-auth.js.org/providers/email

You could use that for user signup, and after that provide the option to connect any account with the same email address.

**Reply by Gregoor** [reply]
Oh I forgot to mention, our Login providers are Google. Isn't account linking not-yet-supported? Or is there another means of connecting account that I'm missing right now?
Maybe a note: this is blocking login from working, as they would only be able to login with the session-link, so there is no active session yet.

**Reply by balazsorban44** [reply]
account linking is and has always been supported. the only requirement is that you can only connect accounts with the same email address.

it's an automatic process. when your user has logged in with the email provider, you can present them with a link Google account button, which calls `signIn("google")` and the account should be connected to the user automatically

**Reply by kcrtr** [reply]
I think what @balazsorban44 is suggesting ties into what I'm suggesting. However I think an important part of the exact solution hinges on if you want to limit access to the app to only user accounts you've created. Is that the case?

**Reply by Gregoor** [reply]
I think what I'm failing to communicate effectively is that in the scenario which is not working for me the user does not have a session yet. And their only means of signing in/up is through an account which is not connected to a user yet, but a user for that email does already exist. So I basically want to automatically link accounts on sign-in, which from the FAQ I understand to be not supported currently.



> However I think an important part of the exact solution hinges on if you want to limit access to the app to only user accounts you've created. Is that the case?

Not necessarily, there's both invited users and users who signed-up themselves in the system.

**edit:** Thinking about it some more, I think I ultimately just really need account linking, on sign-in, for oauth providers that I trust. So I guess best-case I could configure to be trusted, and have them be auto-linked (just like email is auto-trusted by default). Would you be open for PRs that enable auto-linking-on-signin-through-explicit-opt-in @balazsorban44 ?

**Reply by balazsorban44** [reply]
No, it is not supported to link accounts automatically without the user being logged in, and I don't see it being added at any time. As I haven't written that piece of code myself, let me link to the comments from the source code, which explains beautifully why it is not allowed:

https://github.com/nextauthjs/next-auth/blob/ba39efb2566466dfe695bb90513050896d96a2d2/src/core/lib/callback-handler.ts#L163-L179

**Reply by Gregoor** [reply]
Thanks, that does make sense to me in general. I think in our case, we would be open to relax it for certain providers, which we have a high degree of trust in, but I do appreciate the principled security view on it.

I am thinking now that thus I should probably just create these temporary-users without emails. Then people won't run into the `AccountNotLinkedError` and will be able to log-in. But it will not be the account that was created for them, so we will have to think about merging of users.

**Reply by Gregoor** [reply]
We've done some thinking and implementing and ended up just exposing a form on the error page for `OAuthAccountNotLinked`, which lets people send themselves a recovery-email (includes a link with a JWT in the URL, which gets put into the cookie, as mentioned in the first post).
Now it would be neat to have information about the account for which auth failed, to prefill the form with the user's email address. It looks like there is no event for errors during auth? Something like that would be very useful for us, e.g.:
```ts
NextAuth({
  ...,
  events: {
    async signUpError({ error, res, account }) {
      if (error.name == "OAuthAccountNotLinked") {
        res.writeHead(HttpStatus.FOUND, { Location: `/auth/error?error=OAuthAccountNotLinked&email=${account.email}` }).end();
      }
    }
  },
  ...
})
```

**Reply by Gregoor** [reply]
@balazsorban44 hope it's alright I ping you. I'm curious if adding an error event to next-auth, which includes the account for which it errord, would be an option? I'd be happy to send in a PR
